python36Packages.ipython: 7.17 -> 7.16.1 (downgrade on Python 3.6 for compatibility)

kini opened the pull request:
###### Motivation for this change

IPython 7.17 and higher don't support Python 3.6 anymore.

    nix-repl> :b python36.withPackages (ps: [ ps.ipython ])
    builder for '/nix/store/q8v4f89xwv35a3idb9z345z6n3nzfycb-python3.6-ipython-7.17.0.drv' failed with exit code 1; last 10 log lines:
      Python 3.5 was supported with IPython 7.0 to 7.9.
      Python 3.6 was supported with IPython up to 7.16.

      See IPython `README.rst` file for more information:

          https://github.com/ipython/ipython/blob/master/README.rst

      Python sys.version_info(major=3, minor=6, micro=12, releaselevel='final', serial=0) detected.
    cannot build derivation '/nix/store/mhpdarp18z6skzswrl7sbgzv8hr4gwih-python3-3.6.12-env.drv': 1 dependencies couldn't be built
    [0 built (1 failed)]
    error: build of '/nix/store/mhpdarp18z6skzswrl7sbgzv8hr4gwih-python3-3.6.12-env.drv' failed

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

kini committed:
python36Packages.ipython: 7.17 -> 7.16.1 (downgrade)

IPython versions 7.17 and higher don't support Python 3.6 anymore, so
this commit adds back the old expression for 7.16.1 from before
ba1038a98bc2e3aa399c5ef3c4b03d42c94e1bea for use in python36Packages
only.

    nix-repl> :b python36.withPackages (ps: [ ps.ipython ])
    builder for '/nix/store/q8v4f89xwv35a3idb9z345z6n3nzfycb-python3.6-ipython-7.17.0.drv' failed with exit code 1; last 10 log lines:
      Python 3.5 was supported with IPython 7.0 to 7.9.
      Python 3.6 was supported with IPython up to 7.16.

      See IPython `README.rst` file for more information:

          https://github.com/ipython/ipython/blob/master/README.rst

      Python sys.version_info(major=3, minor=6, micro=12, releaselevel='final', serial=0) detected.
    cannot build derivation '/nix/store/mhpdarp18z6skzswrl7sbgzv8hr4gwih-python3-3.6.12-env.drv': 1 dependencies couldn't be built
    [0 built (1 failed)]
    error: build of '/nix/store/mhpdarp18z6skzswrl7sbgzv8hr4gwih-python3-3.6.12-env.drv' failed

jonringer reviewed (commented):
To comply with [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes) please have the commit message name be of the format
```
<pkg-name>: <subject-line>
```
for more examples, please look at https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes

in your case, the commit message should be:
```
python36Packages: disable
```

cc @FRidh

kini commented:
Argh, sorry about that. I've now changed the commit message to 

```
python36Packages.ipython: python36 not supported
```

to match your commit ec1dc42420b1640007a2a877c122cbbb23226a45 from last year.

FRidh commented:
Thanks. We should probably duplicate the expression for 7.16 for Python 3.6.

kini commented:
Makes sense. Should we also have IPython 7.9 for compatibility with Python 3.5? It looks like we never actually had IPython 7.9, since we jumped from 7.8 to 7.10 in 1f7ee3e040e1efd2e9803356b1c71f03c688afda.

FRidh commented:
3.5 has been dropped from Nixpkgs

kini commented:
Ah, I didn't notice that, thanks. (I still see `isPy35` defined in `pkgs/top-level/python-packages.nix` and used in various places, but I guess it's just always false then.)

I've added a duplicate expression for 7.16 and plugged it into `python36Packages`, and changed the commit message accordingly.

drewrisinger commented on pkgs/top-level/python-packages.nix:
```diff
@@ -2908,6 +2908,8 @@ in {
 
   ipython = if isPy27 then
     callPackage ../development/python-modules/ipython/5.nix { }
+  else if isPy36 then
+    callPackage ../development/python-modules/ipython/7.16.nix { }
```
I'd be cautious & thoroughly test this if-elseif-else syntax in Nix. I've had troubles before where it doesn't evaluate as you would expect from other languages.

kini replied:
I had the same concern, so I tested it by building `ipython` for Python 2.7, 3.6, and 3.7:

```
nix-repl> :b python27Packages.ipython 

this derivation produced the following outputs:
  out -> /nix/store/2ihnjhkzr2mj1in0l6wkm21xar4pvc2c-python2.7-ipython-5.8.0

nix-repl> :b python36Packages.ipython 

this derivation produced the following outputs:
  out -> /nix/store/a0i7j61c7vd33kwbkbg5jwqnla1qx0zm-python3.6-ipython-7.16.1

nix-repl> :b python37Packages.ipython 

this derivation produced the following outputs:
  out -> /nix/store/qdb4fjicn2xhp8f6d3brg33n45zfm7lx-python3.7-ipython-7.17.0
```

You can see from the pathnames that the correct `ipython` version numbers are getting selected for each Python version.
